rust/arrow-pyarrow-integration-testing/tests/test_sql.py

```diff
@@ -78,3 +78,23 @@ def test_time32_python(self):
         del expected
         # No leak of C++ memory
         self.assertEqual(old_allocated, pyarrow.total_allocated_bytes())
+
+    def test_list_array(self):
+        """
+        Python -> Rust -> Python
+        """
+        old_allocated = pyarrow.total_allocated_bytes()
+        a = pyarrow.array([[1, 2], [1, 2]], pyarrow.list_(pyarrow.int64()))
+        b = arrow_pyarrow_integration_testing.round_trip(a)
+        # list equality does not work, so we check the elements
```

Comment:
@ritchie46 is this a `pyarrow` or Rust limitation?

Reply:
This is `pyarrow`.

Reply:
Interesting. Equality is not part of the spec but it is very important that we get this right. Let's cc @pitrou here to see if this is something on rust or C++ side.

Reply:
When you say "does not work", what message do you get?

Reply:
Sorry, it could be I meant it evaluated to false where they were equal in values.

I am on mobile atm. I could check that later.

Reply:
Ok, I can confirm that `pyarrow` list equality works as expected. 

The reason why list equality evaluated to false even though the values were equal is due to a slight type difference. After the round trip the type has an additional `not null` flag.

The not null indicatior is correct in this case. Is this problematic for this PR @jorgecarleitao ?

Reply:
Just in case, I set the nullable flag to [true](https://github.com/apache/arrow/pull/9778/commits/90d68d89abb2f2606005162199ab13330a909852). I don't know if its ever used, though?

Reply:
I don't think the nullable flag is very much currently.
When you say "it did not have an effect on the test", what do you mean precisely? The test should certainly succeed.

As a first step, can you add the following code:
```python
b.validate(full=True)
assert a.to_pylist() == b.to_pylist()
```

Also, you should make the example slightly more "interesting", instead of `[[1, 2], [1, 2]]` use e.g. `[[], None, [1, 2], [4, 5, 6]]`.

Reply:
> As a first step, can you add the following code:
> 
> ```python
> b.validate(full=True)
> assert a.to_pylist() == b.to_pylist()
> ```
> 
> Also, you should make the example slightly more "interesting", instead of `[[1, 2], [1, 2]]` use e.g. `[[], None, [1, 2], [4, 5, 6]]`.


I did add those test you recommended. They run successfully. What does fail, is testing equality with `a == b`. Which I assume is due to the difference in the type `is_nullable` flag.

Reply:
> When you say "it did not have an effect on the test", what do you mean precisely? The test should certainly succeed.

And what I mean by "it did not have an effect on the test", is that adding [the `is_nullable` flag](https://github.com/apache/arrow/pull/9778/commits/90d68d89abb2f2606005162199ab13330a909852), did not lead to `a == b` evaluating to `True`.

Reply:
Well, then you can also add `assert a.type == b.type` to check if the nullable flag is correctly set.

Reply:
> Well, then you can also add `assert a.type == b.type` to check if the nullable flag is correctly set.

This fails. So my question is, is this due to the code in this PR (i.e. on the Rust side). If so, than I need a bit of guidance of where this should be defined.

And more general, what is the use of this flag?

Reply:
By the [format spec](https://arrow.apache.org/docs/format/Columnar.html#schema-message):
> Whether the field is semantically nullable. While this has no bearing on the array’s physical layout, many systems distinguish nullable and non-nullable fields and we want to allow them to preserve this metadata to enable faithful schema round trips.

So, while it's not very useful in itself for Arrow implementations, it's a piece of metadata that's nice to preserve.

Reply:
> `list<item: int64> list<: int64 not null>`

Ah, besides the nullable bit being unset, you also see that in the second case, the name of the nested field inside the list type is empty (it's "item" in the first case).